File: frontends/plotly/src/protocol.ts
export interface Configuration {
  hz: number;
  points: number;
  append_count: number;
  width: number;
  height: number;
  waveform_mode: 'replace' | 'append';
  image_mode: 'scalar' | 'rgb';
  view: 'waveform' | 'image' | 'both';
  seed: number;
  generation: number;
}

export interface Frame {
  seq: number;
  generation: number;
  emitted_at_ms: number;
  config: Configuration;
  waveform?: Float32Array;
  image?: Float32Array | Uint8Array;
}

function object(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('Expected a JSON object');
  }
  return value as Record<string, unknown>;
}

function integer(value: unknown, label: string, minimum = 0): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < minimum) {
    throw new Error(`Invalid ${label}`);
  }
  return value;
}

export function parseConfiguration(value: unknown): Configuration {
  const config = object(value);
  if (typeof config.hz !== 'number' || !Number.isFinite(config.hz) || config.hz <= 0 || config.hz > 120) {
    throw new Error('Source frequency must be in (0, 120] Hz');
  }
  const points = integer(config.points, 'points', 1);
  const append_count = integer(config.append_count, 'append_count', 1);
  if (append_count > points) throw new Error('append_count exceeds points');
  if (config.waveform_mode !== 'append' && config.waveform_mode !== 'replace') {
    throw new Error('Invalid waveform_mode');
  }
  if (config.image_mode !== 'scalar' && config.image_mode !== 'rgb') {
    throw new Error('Invalid image_mode');
  }
  if (!['waveform', 'image', 'both'].includes(String(config.view))) throw new Error('Invalid view');
  return {
    hz: config.hz, points, append_count,
    width: integer(config.width, 'width', 1),
    height: integer(config.height, 'height', 1),
    waveform_mode: config.waveform_mode,
    image_mode: config.image_mode,
    view: config.view as Configuration['view'],
    seed: integer(config.seed, 'seed'),
    generation: integer(config.generation, 'generation'),
  };
}

export function parseFrame(buffer: ArrayBuffer): Frame {
  if (buffer.byteLength < 4) throw new Error('Truncated frame prefix');
  const view = new DataView(buffer);
  const headerLength = view.getUint32(0, true);
  const payloadStart = Math.ceil((4 + headerLength) / 4) * 4;
  if (headerLength === 0 || payloadStart > buffer.byteLength) throw new Error('Truncated frame header');
  const header = object(JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(
    new Uint8Array(buffer, 4, headerLength),
  )));
  if (header.version !== 1) throw new Error('Unsupported protocol version');
  const config = parseConfiguration(header.config);
  if (typeof header.emitted_at_ms !== 'number' || !Number.isFinite(header.emitted_at_ms)) {
    throw new Error('Invalid emission timestamp');
  }
  const frame: Frame = {
    seq: integer(header.seq, 'seq'),
    generation: integer(header.generation, 'generation'),
    emitted_at_ms: header.emitted_at_ms,
    config,
  };
  if (frame.generation !== config.generation) throw new Error('Inconsistent frame generation');
  if (!Array.isArray(header.arrays)) throw new Error('Missing array descriptors');
  const occupied: [number, number][] = [];
  for (const rawDescriptor of header.arrays) {
    const descriptor = object(rawDescriptor);
    const name = descriptor.name;
    if (name !== 'waveform' && name !== 'image') throw new Error('Unknown array name');
    if (frame[name]) throw new Error(`Duplicate ${name} array`);
    const expectedShape = name === 'waveform'
      ? [config.points]
      : config.image_mode === 'rgb' ? [config.height, config.width, 3] : [config.height, config.width];
    const expectedDtype = name === 'image' && config.image_mode === 'rgb' ? 'uint8' : 'float32';
    if (descriptor.dtype !== expectedDtype) throw new Error(`Invalid ${name} dtype`);
    if (!Array.isArray(descriptor.shape) || descriptor.shape.length !== expectedShape.length ||
        descriptor.shape.some((size, index) => size !== expectedShape[index])) {
      throw new Error(`Invalid ${name} shape`);
    }
    const count = expectedShape.reduce((product, size) => product * size, 1);
    const offset = integer(descriptor.offset, 'offset');
    const length = integer(descriptor.nbytes, 'nbytes');
    const expectedBytes = count * (expectedDtype === 'float32' ? 4 : 1);
    if (!Number.isSafeInteger(expectedBytes) || length !== expectedBytes ||
        offset + length > buffer.byteLength - payloadStart) throw new Error(`Truncated ${name} payload`);
    if (expectedDtype === 'float32' && offset % 4 !== 0) throw new Error('Misaligned float32 payload');
    if (occupied.some(([start, end]) => offset < end && offset + length > start)) {
      throw new Error('Overlapping array payloads');
    }
    occupied.push([offset, offset + length]);
    if (name === 'waveform') frame.waveform = new Float32Array(buffer, payloadStart + offset, count);
    else frame.image = expectedDtype === 'float32'
      ? new Float32Array(buffer, payloadStart + offset, count)
      : new Uint8Array(buffer, payloadStart + offset, count);
  }
  if (config.view !== 'image' && !frame.waveform) throw new Error('Missing waveform payload');
  if (config.view !== 'waveform' && !frame.image) throw new Error('Missing image payload');
  return frame;
}

/** Return transport credit after offering the decoded frame, before deferred plotting starts. */
export function acceptStreamPacket(buffer: ArrayBuffer, offer: (frame: Frame) => void,
  acknowledge: (message: string) => void): void {
  const frame = parseFrame(buffer);
  offer(frame);
  acknowledge(JSON.stringify({ ack: frame.seq, generation: frame.generation }));
}

export function parseReplay(buffer: ArrayBuffer): Frame[] {
  if (buffer.byteLength < 4) throw new Error('Truncated replay prefix');
  const view = new DataView(buffer);
  const count = view.getUint32(0, true);
  if (count === 0 || count > Math.floor((buffer.byteLength - 4) / 8)) {
    throw new Error('Invalid replay frame count');
  }
  let cursor = 4;
  const frames: Frame[] = [];
  for (let index = 0; index < count; index += 1) {
    if (cursor + 4 > buffer.byteLength) throw new Error('Truncated replay length');
    const length = view.getUint32(cursor, true);
    cursor += 4;
    if (length === 0 || cursor + length > buffer.byteLength) throw new Error('Truncated replay packet');
    frames.push(parseFrame(buffer.slice(cursor, cursor + length)));
    cursor += length;
  }
  if (cursor !== buffer.byteLength) throw new Error('Trailing replay bytes');
  return frames;
}

File: frontends/jfreechart/src/main/java/org/plotbench/Protocol.java
package org.plotbench;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.*;

/** Strict protocol decoding. Array slices remain immutable and owned by the packet. */
final class Protocol {
  static final int MAX_REPLAY = 256 * 1024 * 1024;
  static final int MAX_PACKET = MAX_REPLAY + 65540;
  static final ObjectMapper JSON =
      new ObjectMapper().enable(JsonParser.Feature.STRICT_DUPLICATE_DETECTION);

  record Config(
      double hz,
      int points,
      int appendCount,
      int curves,
      int waveformPlots,
      int width,
      int height,
      int imagePlots,
      String waveformMode,
      String imageMode,
      String view,
      long seed,
      long generation) {}

  record Frame(
      long seq, long generation, double emitted, Config config, Map<String, ByteBuffer> arrays) {
    Frame withSequence(long sequence) {
      return new Frame(sequence, generation, emitted, config, arrays);
    }
  }

  static long integer(JsonNode node, String key, long min, long max) {
    JsonNode v = node.get(key);
    if (v == null
        || !v.isIntegralNumber()
        || !v.canConvertToLong()
        || v.longValue() < min
        || v.longValue() > max) throw bad("invalid " + key);
    return v.longValue();
  }

  static double number(JsonNode node, String key) {
    JsonNode v = node.get(key);
    if (v == null || !v.isNumber() || !Double.isFinite(v.doubleValue()))
      throw bad("invalid " + key);
    return v.doubleValue();
  }

  static String choice(JsonNode node, String key, String... choices) {
    JsonNode v = node.get(key);
    if (v == null || !v.isTextual() || !Arrays.asList(choices).contains(v.textValue()))
      throw bad("invalid " + key);
    return v.textValue();
  }

  static IllegalArgumentException bad(String msg) {
    return new IllegalArgumentException(msg);
  }

  static Frame decode(byte[] data) {
    return decode(ByteBuffer.wrap(data));
  }

  static Frame decode(ByteBuffer input) {
    ByteBuffer data = input.slice().asReadOnlyBuffer().order(ByteOrder.LITTLE_ENDIAN);
    if (data.remaining() < 4 || data.remaining() > MAX_PACKET) throw bad("invalid packet length");
    int n = data.getInt();
    if (n < 1 || n > 65536 || ((4L + n + 3) & ~3L) > data.limit())
      throw bad("invalid header length");
    byte[] header = new byte[n];
    data.get(header);
    JsonNode h;
    try {
      h = JSON.readTree(header);
    } catch (IOException ex) {
      throw bad("invalid JSON header");
    }
    if (h == null || !h.isObject()) throw bad("invalid header");
    if (!h.path("version").isIntegralNumber()
        || !h.path("version").canConvertToInt()
        || h.path("version").intValue() != 2)
      throw bad("unsupported protocol version: " + h.path("version") + "; expected 2");
    long seq = integer(h, "seq", 0, Long.MAX_VALUE);
    long generation = integer(h, "generation", 0, Long.MAX_VALUE);
    double emitted = number(h, "emitted_at_ms");
    JsonNode c = h.path("config");
    double hz = number(c, "hz");
    if (hz <= 0 || hz > 120) throw bad("invalid hz");
    int points = (int) integer(c, "points", 1, 10_000_000);
    Config config =
        new Config(
            hz,
            points,
            (int) integer(c, "append_count", 1, points),
            (int) integer(c, "curves", 1, 64),
            (int) integer(c, "waveform_plots", 1, 16),
            (int) integer(c, "width", 1, 8192),
            (int) integer(c, "height", 1, 8192),
            (int) integer(c, "image_plots", 1, 16),
            choice(c, "waveform_mode", "append", "replace"),
            choice(c, "image_mode", "scalar", "rgb"),
            choice(c, "view", "both", "image", "waveform"),
            integer(c, "seed", 0, Long.MAX_VALUE),
            integer(c, "generation", 0, Long.MAX_VALUE));
    if (config.generation != generation) throw bad("generation mismatch");
    int payload = (4 + n + 3) & ~3;
    for (int i = 4 + n; i < payload; i++) if (data.get(i) != 0) throw bad("nonzero padding");
    JsonNode descriptors = h.path("arrays");
    if (!descriptors.isArray()) throw bad("missing arrays");
    Map<String, ByteBuffer> arrays = new HashMap<>();
    long offset = 0;
    for (JsonNode a : descriptors) {
      String name = choice(a, "name", "waveform", "image");
      if (arrays.containsKey(name)) throw bad("duplicate array");
      int[] shape;
      int item = 4;
      if (name.equals("waveform")) {
        if (config.view.equals("image")) throw bad("unexpected waveform");
        shape = new int[] {config.waveformPlots, config.curves, points};
      } else {
        if (config.view.equals("waveform")) throw bad("unexpected image");
        shape =
            config.imageMode.equals("rgb")
                ? new int[] {config.imagePlots, config.height, config.width, 3}
                : new int[] {config.imagePlots, config.height, config.width};
        if (config.imageMode.equals("rgb")) item = 1;
      }
      choice(a, "dtype", item == 1 ? "uint8" : "float32");
      if (!a.path("shape").isArray() || a.path("shape").size() != shape.length)
        throw bad("shape mismatch");
      long bytes = item;
      for (int i = 0; i < shape.length; i++) {
        JsonNode d = a.path("shape").get(i);
        if (!d.isIntegralNumber() || d.longValue() != shape[i]) throw bad("shape mismatch");
        bytes *= shape[i];
      }
      integer(a, "offset", offset, offset);
      integer(a, "nbytes", bytes, bytes);
      if (offset % item != 0 || bytes > data.limit() - payload - offset)
        throw bad("truncated array");
      arrays.put(
          name,
          data.slice(payload + (int) offset, (int) bytes)
              .asReadOnlyBuffer()
              .order(ByteOrder.LITTLE_ENDIAN));
      offset += bytes;
    }
    if (arrays.size() != (config.view.equals("both") ? 2 : 1)
        || offset != data.limit() - payload
        || offset > MAX_REPLAY) throw bad("missing arrays or trailing bytes");
    return new Frame(seq, generation, emitted, config, Collections.unmodifiableMap(arrays));
  }

  static List<Frame> replay(byte[] bytes) {
    if (bytes.length < 4 || bytes.length > MAX_REPLAY) throw bad("invalid replay size");
    ByteBuffer b = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
    int n = b.getInt();
    if (n < 2 || n > 256) throw bad("invalid replay count");
    List<Frame> frames = new ArrayList<>();
    for (int i = 0; i < n; i++) {
      if (b.remaining() < 4) throw bad("truncated replay length");
      int len = b.getInt();
      if (len < 0 || len > b.remaining()) throw bad("truncated replay frame");
      Frame f = decode(b.slice(b.position(), len));
      b.position(b.position() + len);
      if (i > 0 && (!f.config.equals(frames.get(0).config) || f.seq <= frames.get(i - 1).seq))
        throw bad("inconsistent replay");
      frames.add(f);
    }
    if (b.hasRemaining()) throw bad("trailing replay bytes");
    return List.copyOf(frames);
  }

  static int[] palette(byte[] bytes) throws IOException {
    JsonNode lut = JSON.readTree(bytes);
    if (lut == null || !lut.isArray() || lut.size() != 256) throw bad("invalid palette");
    int[] colors = new int[256];
    for (int i = 0; i < 256; i++) {
      JsonNode rgb = lut.get(i);
      if (!rgb.isArray() || rgb.size() != 3) throw bad("invalid palette entry");
      int color = 0xff000000;
      for (int j = 0; j < 3; j++) {
        JsonNode v = rgb.get(j);
        if (!v.isIntegralNumber() || !v.canConvertToInt() || v.intValue() < 0 || v.intValue() > 255)
          throw bad("invalid color");
        color |= v.intValue() << (16 - 8 * j);
      }
      colors[i] = color;
    }
    return colors;
  }
}
